modules/translation/i18n/i18n.go

```diff
@@ -85,114 +103,194 @@ func (ls *LocaleStore) reloadLocaleByIni(langName string, source interface{}) er
+		return
+	}
+
+	l.reloadMu.RUnlock()
+	l.reloadMu.Lock() // (NOTE: a pre-emption can occur between these two locks so we need to recheck)
+
+	if now.Sub(l.lastReloadCheckTime) < time.Second || l.sourceFileInfo == nil || l.sourceFileName == "" {
+		l.reloadMu.Unlock()
+		l.reloadMu.RLock()
+		return
+	}
+
+	l.lastReloadCheckTime = now
+	sourceFileInfo, err := os.Stat(l.sourceFileName)
+	if err != nil || sourceFileInfo.ModTime().Equal(l.sourceFileInfo.ModTime()) {
+		l.reloadMu.Unlock()
+		l.reloadMu.RLock()
+		return
+	}
+
+	l.store.reloadMu.Lock()
+	err = l.store.reloadLocaleByIni(l.langName, l.sourceFileName)
+	l.store.reloadMu.Unlock()
+
+	if err == nil {
+		l.sourceFileInfo = sourceFileInfo
+	} else {
+		log.Error("unable to live-reload the locale file %q, err: %v", l.sourceFileName, err)
+	}
+
+	l.reloadMu.Unlock()
+	l.reloadMu.RLock()
+}
+
 // Tr translates content to locale language. fall back to default language.
 func (l *locale) Tr(trKey string, trArgs ...interface{}) string {
-	if l.store.reloadMu != nil {
-		l.store.reloadMu.Lock()
-		defer l.store.reloadMu.Unlock()
-		now := time.Now()
-		if now.Sub(l.lastReloadCheckTime) >= time.Second && l.sourceFileInfo != nil && l.sourceFileName != "" {
-			l.lastReloadCheckTime = now
-			if sourceFileInfo, err := os.Stat(l.sourceFileName); err == nil && !sourceFileInfo.ModTime().Equal(l.sourceFileInfo.ModTime()) {
-				if err = l.store.reloadLocaleByIni(l.langName, l.sourceFileName); err == nil {
-					l.sourceFileInfo = sourceFileInfo
-				} else {
-					log.Error("unable to live-reload the locale file %q, err: %v", l.sourceFileName, err)
-				}
-			}
-		}
+	if l.reloadMu != nil {
+		l.reloadMu.RLock()
+		defer l.reloadMu.RUnlock()
+		l.reloadIfNeeded()
 	}
+
 	msg, _ := l.tryTr(trKey, trArgs...)
 	return msg
 }
 
 func (l *locale) tryTr(trKey string, trArgs ...interface{}) (msg string, found bool) {
 	trMsg := trKey
-	textIdx, ok := l.store.textIdxMap[trKey]
+
+	if l.store.reloadMu != nil {
+		l.store.reloadMu.RLock()
+	}
+	idx, ok := l.store.trKeyToIdxMap[trKey]
+	if l.store.reloadMu != nil {
+		l.store.reloadMu.RUnlock()
+	}
+
 	if ok {
-		if msg, found = l.textMap[textIdx]; found {
+		if msg, found = l.idxToMsgMap[idx]; found {
 			trMsg = msg // use current translation
 		} else if l.langName != l.store.defaultLang {
-			if def, ok := l.store.localeMap[l.store.defaultLang]; ok {
-				return def.tryTr(trKey, trArgs...)
+
+			// attempt to get the default language from the locale store
+			if l.store.reloadMu != nil {
+				l.store.reloadMu.RLock()
+			}
+			def, ok := l.store.localeMap[l.store.defaultLang]
+			if l.store.reloadMu != nil {
+				l.store.reloadMu.RUnlock()
+			}
+
+			if ok {
+				if def.reloadMu != nil {
+					def.reloadMu.RLock()
+				}
+				if msg, found = def.idxToMsgMap[idx]; found {
+					trMsg = msg // use current translation
+				}
+				if def.reloadMu != nil {
+					def.reloadMu.RUnlock()
+				}
```

Comment:
But it's a very rare case. The locale list seldom changes in months. So I think it's not worth to make it too complex, and it can save some CPU time from locking/unlocking.

And, if the locale list is changed, the setting options should also be changed to make them take affect, otherwise they still can not be loaded.

Reply:
if you mean the existing, it could be check in `tryTr` instead, without more locking.

Reply:
![image](https://user-images.githubusercontent.com/2114189/176146758-58c53c16-45cb-4492-be74-9f5934015250.png)